ldap.go

```diff
@@ -15,87 +18,459 @@ const (
 	LdapConnectionModeTLS                          = iota
 )
 
-type ldapAuthenticator struct {
-	//con *ldap.Conn
+var (
+	ldapDomain string
+)
+
+type LdapAuthenticator struct {
+	con *ldap.LDAPConnection
+}
+
+type ldapUserStore struct {
 	con *ldap.LDAPConnection
+	db  *sql.DB
 }
 
-func (x *ldapAuthenticator) Authenticate(identity, password string) (er error) {
+func (x *LdapAuthenticator) Authenticate(identity, password string) (er error) {
 	if len(password) == 0 {
 		// Many LDAP servers (or AD) will allow an anonymous BIND.
 		// I've never seen the need for a password-less user authenticated against LDAP.
 		er = ErrInvalidPassword
 		return
 	}
-	err := x.con.Bind(identity, password)
+	err := x.con.Bind(fmt.Sprintf(`%v@%v`, identity, ldapDomain), password)
 	if err != nil {
 		er = NewError(ErrIdentityAuthNotFound, err.Error())
 	}
+
 	return
 }
 
-func (x *ldapAuthenticator) Close() {
+func (x *LdapAuthenticator) Close() {
 	if x.con != nil {
 		x.con.Close()
 		x.con = nil
 	}
 }
 
-type ldapUserStore struct {
-	//con *ldap.Conn
-	con *ldap.LDAPConnection
-}
-
-func (x *ldapUserStore) SetPassword(identity, password string) error {
+func (x *ldapUserStore) SetPassword(userId UserId, password string) error {
 	return ErrUnsupported
 }
 
-func (x *ldapUserStore) ResetPasswordStart(identity string, expires time.Time) (string, error) {
+func (x *ldapUserStore) ResetPasswordStart(userId UserId, expires time.Time) (string, error) {
 	return "", ErrUnsupported
 }
 
-func (x *ldapUserStore) ResetPasswordFinish(identity string, token string, password string) error {
+func (x *ldapUserStore) ResetPasswordFinish(userId UserId, token string, password string) error {
 	return ErrUnsupported
 }
 
-func (x *ldapUserStore) CreateIdentity(identity, password string) error {
+func (x *ldapUserStore) CreateIdentity(email, username, firstname, lastname, mobilenumber, password string) (UserId, error) {
 	return ErrUnsupported
 }
 
 func (x *ldapUserStore) RenameIdentity(oldIdent, newIdent string) error {
 	return ErrUnsupported
 }
 
-func (x *ldapUserStore) GetIdentities() ([]string, error) {
-	return []string{}, ErrUnsupported
+func (x *ldapUserStore) GetIdentities() ([]AuthUser, error) {
+	rows, err := x.db.Query(`SELECT userid, email, username, firstname, lastname, mobile FROM authuserstore WHERE (archived = false OR archived IS NULL) ORDER BY username`)
```

Comment:
This is a mega-nasty abstraction failure, but we can deal with all of the cleanup after the demo.